edalize/nextpnr.py

```diff
@@ -0,0 +1,168 @@
+                        "desc" : "Determines if nextpnr is run as a part of bigger toolchain, or as a standalone tool"
+                    },
+                ],
+                "lists" : [
+                    {
+                        "name" : "nextpnr_impl_options",
+                        "type" : "String",
+                        "desc" : "Additional options for the implementation command"
+                    },
+                ]
+            }
+
+            combined_members = nextpnr_help["members"]
+            combined_lists = nextpnr_help.get("lists", [])
+            yosys_members = yosys_help["members"]
+            yosys_lists = yosys_help["lists"]
+
+            combined_members.extend(m for m in yosys_members if m["name"] not in [i["name"] for i in combined_members])
+            combined_lists.extend(l for l in yosys_lists if l["name"] not in [i["name"] for i in combined_lists])
+
+            return {
+                "description" : "Open source Place and Route tool targeting many different FPGAs",
+                "members": combined_members,
+                "lists": combined_lists
+            }
+
+    def configure_main(self):
+        (src_files, incdirs) = self._get_fileset_files()
+
+        yosys_synth_options = self.tool_options.get("yosys_synth_options", "")
+        yosys_template = self.tool_options.get("yosys_template", None)
+        yosys_edam = {
+                "files"         : self.files,
+                "name"          : self.name,
+                "toplevel"      : self.toplevel,
+                "parameters"    : self.parameters,
+                "tool_options"  : {
+                                    "yosys" : {
+                                        "arch" : "xilinx",
+                                        "yosys_synth_options" : yosys_synth_options,
+                                        "yosys_template" : yosys_template,
+                                        "yosys_as_subtool" : True,
+                                    }
+                                }
+                }
+
+        yosys = getattr(import_module("edalize.yosys"), "Yosys")(yosys_edam, self.work_root)
+        yosys.configure()
+
+        part_of_toolchain = self.tool_options.get("nextpnr_as_subtool", False)
+
+        file_table = []
+
+        output_format = self.tool_options.get("output_format", "fasm")
+        arch = self.tool_options.get("arch")
+
+        if not arch in getattr(self, "archs"):
+            logger.error('Missing or invalid "arch" parameter in "tool_options"')
+
+        synth_design = None
+        chipdb = None
+        device = None
+        xdc = None
+
+        for f in src_files:
+            if f.file_type in ["synthJson"]:
+                synth_design = f.name
+            elif f.file_type in ["bba"]:
+                chipdb = f.name
+            elif f.file_type in ["device"]:
+                device = f.name
+            elif f.file_type in ["xdc"]:
```

Comment:
If we want to reuse the nextpnr class in apicula, trellis and icestorm (which I think we should do), then we need to be more flexible here since neither nextpnr_ice40, nextpnr_ecp5 or nextpnr_gowin has any knowledge of chipdb and xdc files. I'm starting to think that we perhaps should either begin with pulling out the nextpnr support from the existing backends and then add the symbiflow stuff on top of that. Or, put all this directly in the symbiflow class and extract nextpnr from all backends later

Reply:
I've merged Nextpnr into SymbiFlow class (`configure_nextpnr` method). I believe we can leave this as is?